flume-ng-channels/flume-file-channel/src/test/java/org/apache/flume/channel/file/TestFileChannelErrorMetrics.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.commons.io.FileUtils;
+import org.apache.flume.ChannelException;
+import org.apache.flume.Transaction;
+import org.apache.flume.channel.file.instrumentation.FileChannelCounter;
+import org.apache.flume.event.EventBuilder;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertFalse;
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+
+public class TestFileChannelErrorMetrics extends TestFileChannelBase {
+
+  public TestFileChannelErrorMetrics() {
+    super(1);
+  }
+
+  /**
+   * This tests multiple successful and failed put and take operations
+   * and checks the values of the channel's counters.
+   */
+  @Test
+  public void testEventTakePutErrorCount() throws Exception {
+    final long usableSpaceRefreshInterval = 1;
+    FileChannel channel = Mockito.spy(createFileChannel());
+    Mockito.when(channel.createLogBuilder()).then(new Answer<Log.Builder>() {
+      @Override
+      public Log.Builder answer(InvocationOnMock invocation) throws Throwable {
+        Log.Builder ret = (Log.Builder) invocation.callRealMethod();
+        ret.setUsableSpaceRefreshInterval(usableSpaceRefreshInterval);
+        return ret;
+      }
+    });
+    channel.start();
+
+    FileChannelCounter channelCounter = channel.getChannelCounter();
+
+    Transaction tx = channel.getTransaction();
+    tx.begin();
+    channel.put(EventBuilder.withBody("test1".getBytes()));
+    channel.put(EventBuilder.withBody("test2".getBytes()));
+    tx.commit();
+    tx.close();
+    assertEquals(2, channelCounter.getEventPutAttemptCount());
+    assertEquals(2, channelCounter.getEventPutSuccessCount());
+    assertEquals(0, channelCounter.getEventPutErrorCount());
+
+    tx = channel.getTransaction();
+    tx.begin();
+    channel.take();
+    tx.commit();
+    tx.close();
+    assertEquals(1, channelCounter.getEventTakeAttemptCount());
+    assertEquals(1, channelCounter.getEventTakeSuccessCount());
+    assertEquals(0, channelCounter.getEventTakeErrorCount());
+
+    FileUtils.deleteDirectory(baseDir);
+    Thread.sleep(2 * usableSpaceRefreshInterval);
+
+    tx = channel.getTransaction();
+    tx.begin();
+    ChannelException putException = null;
+    try {
+      channel.put(EventBuilder.withBody("test".getBytes()));
+    } catch (ChannelException ex) {
+      putException = ex;
+    }
+    assertNotNull(putException);
+    assertTrue(putException.getCause() instanceof IOException);
+    assertEquals(3, channelCounter.getEventPutAttemptCount());
+    assertEquals(2, channelCounter.getEventPutSuccessCount());
+    assertEquals(1, channelCounter.getEventPutErrorCount());
+
+    ChannelException takeException = null;
```

Comment:
Please add comment like:

// This is guaranteed to throw an error if the above put() threw an error.